 src/map/mob.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/map/mob.c b/src/map/mob.c
index 7223e98..6fc8ebc 100644
--- a/src/map/mob.c
+++ b/src/map/mob.c
@@ -1012,6 +1012,10 @@ static int mob_delayspawn(int tid, int64 tick, int id, intptr_t data)
 		}
 		md->spawn_timer = INVALID_TIMER;
 		mob->spawn(md);
+		//Pvp enabled [Cookies]
+		if( md->state.boss ) {
+		map->list[bl->m].flag.pvp = 1;
+		}
 	}
 	return 0;
 }
@@ -1100,6 +1104,10 @@ static int mob_spawn(struct mob_data *md)
 		md->bl.m = md->spawn->m;
 		md->bl.x = md->spawn->x;
 		md->bl.y = md->spawn->y;
+		//Pvp enabled [Cookies]
+		if(md->spawn->state.boss){
+			map->list[md->bl.m].flag.pvp = 1;
+		}  
 
 		if( (md->bl.x == 0 && md->bl.y == 0) || md->spawn->xs || md->spawn->ys ) {
 			//Monster can be spawned on an area.
@@ -2951,6 +2959,11 @@ static int mob_dead(struct mob_data *md, struct block_list *src, int type)
 	if (battle_config.mvp_tomb_enabled && md->spawn->state.boss == BTYPE_MVP && map->list[md->bl.m].flag.notomb != 1)
 		mob->mvptomb_create(md, mvp_sd ? mvp_sd->status.name : NULL, time(NULL));
+	//Pvp enabled [Cookies]
+	if(md->spawn->state.boss)  {
+	map->list[md->bl.m].flag.pvp = 0;
+	}
+
 	if( !rebirth ) {
 		status->change_clear(&md->bl,1);
 		mob->setdelayspawn(md); //Set respawning.
-- 
2.27.0.windows.1

